torchdata/dataloader2/utils/dispatch.py

```diff
@@ -0,0 +1,122 @@
+    r"""
+    Given the graph of DataPipe generated by ``traverse_dps`` function, return the
+    DataPipe instance that is the lowest common ancestor of all non-replicable DataPipes
+    (all parent DataPipes of ``sharding_round_robin_dispatch``)
+
+    Note:
+      - If multiple branches share the same source DataPipe and any branch contains a
+        non-replicable DataPipe, the lowest common ancestor of all branches is returned.
+      - If there is any non-replicable DataPipe in a circular-referenced (sub)graph, the
+        whole (sub)graph is treated as non-replicable and the last DataPipe is returned.
+    """
+    assert len(graph) == 1, "DataPipeGraph should only contain a single output DataPipe"
+
+    def _is_round_robin_sharding(dp: DataPipe) -> bool:
+        return type(dp) == ShardingRoundRobinDispatcher
+
+    dps = list_dps(graph)
+    non_replicable_dps: Set[int] = set()
+    for dp in dps:
+        # Skip when it has been visited
+        if id(dp) in non_replicable_dps:
+            continue
+        if _is_round_robin_sharding(dp):
+            parent_dps = list_dps(traverse_dps(dp))
+            for par_dp in parent_dps:
+                non_replicable_dps.add(id(par_dp))
+
+    root_dp_id = list(graph.keys())[0]
+    root_dp, root_graph = graph[root_dp_id]
+
+    cache: Dict[int, Optional[DataPipe]] = {}
+
+    def helper(root_dp_id, root_dp, root_graph) -> Optional[DataPipe]:  # pyre-ignore
+        if root_dp_id in cache:
+            return cache[root_dp_id]
+        if root_dp_id in non_replicable_dps:
+            cache[root_dp_id] = root_dp
+            return root_dp
+        cache[root_dp_id] = None
+        non_replicable_parents = []
+        for dp_id, (dp, src_graph) in root_graph.items():
+            res = helper(dp_id, dp, src_graph)
+            if res is not None:
+                non_replicable_parents.append(res)
+        # `root_dp` becomes the lowest common ancestor of this branch,
+        # if there are more than one unique non-replicable DataPipe prior to it.
+        if len(non_replicable_parents) > 0:
+            # One unique non-replicable DataPipe
+            if len(non_replicable_parents) == 1 or all(
+                dp == non_replicable_parents[0] for dp in non_replicable_parents
+            ):
+                cache[root_dp_id] = non_replicable_parents[0]
+            # Multiple non-replicable DataPipes
+            else:
+                cache[root_dp_id] = root_dp
+        return cache[root_dp_id]
+
+    return helper(root_dp_id, root_dp, root_graph)
+
+
+def find_replicable_branches(graph: DataPipeGraph) -> List[DataPipe]:
+    r"""
+    Given the graph of DataPipe generated by ``traverse_dps`` function, return the DataPipe
+    instances that don't have ``_DummyIterDataPipe`` (non-replicable DataPipe) in the prior graph.
+    """
+    assert len(graph) == 1, "DataPipeGraph should only contain a single output DataPipe"
```

Comment:
does that mean there should be only 1 connected component in the data pipe graph? curious when will there be more than 1 connected components?

Reply:
This is the end of `DataPipe` graph and we always expect only one output. There can be multiple connected within the graph.

The `graph` returned by `traverse_dps` is in the reversed order. Using the following example:
```mermaid
graph TD;
    A-->B-->C-->D;
    A-->C;
    E-->F-->D;
    D-->output;
```
The returned `graph` is:
```
{id(D): (D, {
    id(C): (C, {
            id(B): (B: {
                id(A}: (A, {})
            }),
            id(A): (A, {}),
        }),
    id(F): (F, {
        id(E): (E, {}),
    }),
})}
```
So, the length for first level of `graph` should always be 1